cirq/circuits/circuit.py

```diff
@@ -1258,6 +1258,60 @@ def _json_dict_(self):
     def _from_json_dict_(cls, moments, device, **kwargs):
         return cls(moments, strategy=InsertStrategy.EARLIEST, device=device)
 
+    def raggedy_add(*circuits: 'cirq.Circuit') -> 'cirq.Circuit':
+        """Concatenates circuits while overlapping them if possible.
+
+        Starts with the first circuit (index 0), then iterates over the other
+        circuits while folding them in. To fold two circuits together, they
+        are placed one after the other and then the second one is moved backward
+        until either just before its operations would collide with operations
+        in the first one or else their ends are equal.
+
+        Beware that this method is *not* associative. For example:
+
+            >>> E = cirq.Circuit(cirq.Moment()) * 10
+            >>> H = cirq.Circuit(cirq.H(cirq.LineQubit(0)))
+            >>> f = cirq.Circuit.raggedy_add
+            >>> len(f(E, f(H, H))
+            10
+            >>> len(f(f(E, H), H))
+            11
+
+        Args:
+            circuits: The circuits to concatenate.
+
+        Returns:
+            The concatenated and overlapped circuit.
+        """
+        if len(circuits) == 0:
+            return Circuit()
+        acc = Circuit(circuits[0])
+        for k in range(1, len(circuits)):
+            acc = _raggedy_add_helper(acc, circuits[k])
```

Comment:
Looks like this is O(n^2). Could it be O(n) if the helper mutated acc directly? No biggie if this is just intended for small `n`.

Reply:
Fixed. The runtime is now O(n). In fact it is O(min(n, q * t)) where q is the number of qubits in the system and t is the amount the circuits were overlapped.

Reply:
I'm still seeing square time. In the helper function it constructs the return value from scratch. So if this is called to add, say, N identical single-moment circuits, it would call that helper N times, and constructing that return value is an O(N) operation itself. I may be missing something though.

Reply:
Oh, I see the quadratic time you were talking about now. I thought you were referring to the overlap collision computation (which did used to be quadratic).

The whole thing is now linear time.